### ISSUE TITLE
Inventory updates launched as job's dependency should wait for project updates from project to finish

### ISSUE THREAD

**Issue by AlanCoding** [ansible/awx#11336]
##### ISSUE TYPE
 - Feature Idea

##### SUMMARY
Scenario:

 - Create a project with `update_on_launch` set to true
 - Create an inventory and an SCM inventory source (using project from 1st step) with `update_on_launch` set to true
 - Create a job template using project from 1st step and inventory from 2nd step
 - Launch the job template

When this happens, we will have 3 unified jobs. 1 unified job is the job itself, from the manual launch. There are 2 dependent jobs, the project update and the inventory update.

In this case, the inventory update should be considered to be blocked by the project update. This will assure that it has the latest revision. Right now it does not. Illustration:

```
   ┌───────────┐
   │  spawned  │
   │  project  │◄─────────────────┐
   │  update   │                  │
   └───────────┘              ┌───┴───┐
          ▲                   │       │
          │                   │  job  │
    ┌─────┴─────┐             │       │
    │ spawned   │             └───┬───┘
    │ inventory │                 │
    │ update    │◄────────────────┘
    └───────────┘
```

No blocking should occur if the SCM inventory source uses a project different from the job's project.

There are several other issues that have suggested this behavior should be the case. I wanted to file this specifically as a change request to the task manager dependency and blocking logic.

**Comment by chrismeyersfsu** [comment]
Related issue, border line duplicate? https://github.com/ansible/awx/issues/11233

**Comment by AlanCoding** [comment]
yeah, same issue

**Comment by AlanCoding** [comment]
I have also been wondering, if we cover the important use cases with this improvement, could we remove (or deprecate) the confusing `update_on_project_update` InventorySource field / feature? Ping @ffirg 

https://github.com/ansible/awx/blob/be13a11dd53e9097051612eb58150a44158a54f5/awx/main/models/inventory.py#L989-L991

This actually doesn't work with our current capacity model. Project updates will run inventory updates as a side effect, but they don't have access to the proper capacity filling algorithm when they do this. Starting with receptor mesh (or openshift before then), project updates and inventory updates will run on different nodes. Given that, it's really shaky to have project updates submit inventory updates without the task manager's supervision.

That could create some turbulence, but it would pay down a lot of debt, and could be well worth it.

**Comment by ffirg** [comment]
This seems like a real, validated issue so we should do this. I think we should remove update_on_project_update as there's a fair degree of functional crossover after implementing the logic change. Let's do that unless there's any objections?

**Comment by shanemcd** [comment]
Is this the same as https://github.com/ansible/awx/issues/5692 ?

**Comment by AlanCoding** [comment]
The title of #5692 is overly-broad, but the text is concerned with overlap with the `branch_override` feature. Respecting the overridden branch in inventory updates is expected to be controversial, and low priority in my book. I expect that the issue has attracted attention as users are assuming the broader scope.

This particular issue is outlining a very conservative plan, only concerned with inventory & JT using the same project, and the order of things in that case. This probably doesn't go far enough.

We are trending in the direction of a more aggressive fix to the task manager dependency tracking, fully replacing the `update_on_project_update` functionality. There's been an ask to have _project_ `update_on_launch` be respected for SCM imports in general. This has not been written up yet, and it would fully replace this issue.

**Comment by AlanCoding** [comment]
This is too narrow for what people want, closing in favor of others.

**Comment by Sispheor** [comment]
Hello there.
I'm coming after the battle as it has been already merged in dev.
But it seems that this is a regression. Who can do more can do less. The flag 'update_on_project_update' was very useful to keep the control on what we have in AWX and it was not confusing at all. 
We'll now be full of useless logs about the sync on each job execution.
I guess that we'll now have to code something to execute the sync manually without both flag "update_on_project_update" and "update_on_launch".
From what I can see in the code, the feature has been removed to only keep the "update_on_launch" which a different use-case.

Or maybe I'm missing something. The inventory source will be automatically updated on project update?
Like if the flag was always true?
Because it should be linked, it's a non sense to update a project without updating the inventory source, Isn't ?

**Comment by AlanCoding** [comment]
> Or maybe I'm missing something. The inventory source will be automatically updated on project update?

Almost that. If you launch a job that uses an inventory that updates on launch, and it sources from a project that updates on launch, then the new expectation is that the project will update _then_ the inventory will update with the updated project state and _then_ the job will run with the updated inventory state. This was the primary use case of the old "update_on_project_update" but I'm interested to here if there are any shortcomings that we didn't think about.

**Comment by Sispheor** [comment]
Ok so to keep the control on what is executed we need to uncheck "update_on_launch" from the project and then only start a sync of both project and inventory on demand.

**Comment by zasca** [comment]
I agree with @Sispheor. Not all the time when job template starts need to update project and inventory source. Let's say job template triggered dayli or even hourly while project/inventory updated monthly.
